js/style/style.js

```diff
@@ -385,68 +342,95 @@ class Style extends Evented {
      * ID `before`, or appended if `before` is omitted.
      * @param {StyleLayer|Object} layer
      * @param {string=} before  ID of an existing layer to insert before
-     * @returns {Style} `this`
-     * @private
      */
-    addLayer(layer, before, options) {
+    addLayer(layerObject, before, options) {
         this._checkLoaded();
 
-        if (!(layer instanceof StyleLayer)) {
-            // this layer is not in the style.layers array, so we pass an impossible array index
-            if (this._validate(validateStyle.layer,
-                    `layers.${layer.id}`, layer, {arrayIndex: -1}, options)) return this;
+        const id = layerObject.id;
 
-            layer = StyleLayer.create(layer);
-        }
+        // this layer is not in the style.layers array, so we pass an impossible array index
+        if (this._validate(validateStyle.layer,
+                `layers.${id}`, layerObject, {arrayIndex: -1}, options)) return;
+
+        const layer = StyleLayer.create(layerObject);
         this._validateLayer(layer);
 
-        layer.setEventedParent(this, {layer: {id: layer.id}});
+        layer.setEventedParent(this, {layer: {id: id}});
 
-        this._layers[layer.id] = layer;
-        this._order.splice(before ? this._order.indexOf(before) : Infinity, 0, layer.id);
+        const index = before ? this._order.indexOf(before) : this._order.length;
+        this._order.splice(index, 0, id);
 
-        this._updates.allLayers = true;
-        if (layer.source) {
-            this._updates.sources[layer.source] = true;
+        this._layers[id] = layer;
+
+        delete this._removedLayers[id];
+        this._updateLayer(layer);
+
+        if (layer.type === 'symbol') {
+            this._updatedSymbolOrder = true;
         }
 
-        return this.updateClasses(layer.id);
+        this.updateClasses(id);
+    }
+
+    /**
+     * Add a layer to the map style. The layer will be inserted before the layer with
+     * ID `before`, or appended if `before` is omitted.
+     * @param {StyleLayer|Object} layer
+     * @param {string=} before  ID of an existing layer to insert before
+     */
+    moveLayer(id, before) {
+        this._checkLoaded();
+
+        const layer = this._layers[id];
+        if (!layer) throw new Error(`Layer not found: ${id}`);
+
+        const index = this._order.indexOf(id);
+        this._order.splice(index, 1);
+
+        const newIndex = before ? this._order.indexOf(before) : this._order.length;
+        this._order.splice(newIndex, 0, id);
+
+        if (layer.type === 'symbol') {
+            this._changed = true;
```

Comment:
OK, fixed with https://github.com/mapbox/mapbox-gl-js/pull/3584/commits/6ced8625e1e2e1b2b55d6b385dd5a24b06843962. I also realized that it didn't reset the update state in this case.